src/Data/Bin/Ordering.agda

```diff
@@ -0,0 +1,791 @@
+≤-trans {x} {_} {_} (inj₁ x<y)  (inj₂ y≡z) =  inj₁ (subst (x <_) y≡z x<y)
+≤-trans {_} {_} {z} (inj₂ x≡y) (inj₁ y<z)  =  inj₁ (subst (_< z) (sym x≡y) y<z)
+≤-trans             (inj₂ x≡y) (inj₂ y≡z)  =  inj₂ (trans x≡y y≡z)
+
+≤-antisym : Antisymmetric _≡_ _≤_
+≤-antisym (inj₂ x≡y) _          =  x≡y
+≤-antisym _          (inj₂ y≡x) =  sym y≡x
+≤-antisym (inj₁ x<y) (inj₁ y<x) =  contradiction y<x (<⇒≯ x<y)
+
+------------------------------------------------------------------------------
+-- Several other properties for _<_, _≤_.
+------------------------------------------------------------------------------
+
+≤→≢→< :  ∀ {x y} → x ≤ y → x ≢ y → x < y
+≤→≢→< (inj₁ x<y) _   =  x<y
+≤→≢→< (inj₂ x≡y) x≢y =  contradiction x≡y x≢y
+
+0≤ :  ∀ x → zero ≤ x
+0≤ zero     =  ≤-refl {zero}
+0≤ 2[1+ _ ] =  inj₁ 0<even
+0≤ 1+[2 x ] =  inj₁ 0<odd
+
+≢0⇒> :  ∀ {x} → x ≢ zero → x > zero
+≢0⇒> {zero}     0≢0 =  contradiction refl 0≢0
+≢0⇒> {2[1+ _ ]} _   =  0<even
+≢0⇒> {1+[2 _ ]} _   =  0<odd
+
+≤0⇒≡ :  ∀ {x} → x ≤ zero → x ≡ zero
+≤0⇒≡ {x} x≤0 =  ≤-antisym x≤0 (0≤ x)
+
+0<suc :  ∀ x → zero < suc x
+0<suc x =  ≢0⇒> (suc≢0 {x})
+
+<⇒≱ :  _<_ ⇒ _≱_
+<⇒≱ x<y (inj₁ y<x) =  contradiction y<x (<⇒≯ x<y)
+<⇒≱ x<y (inj₂ y≡x) =  contradiction (sym y≡x) (<⇒≢ x<y)
+
+≤⇒≯ :  _≤_ ⇒ _≯_
+≤⇒≯ {x} {y} x≤y x>y =  <⇒≱ {y} {x} x>y x≤y
+
+≮⇒≥ :  _≮_ ⇒ _≥_
+≮⇒≥ {x} {y} x≮y  with <-cmp x y
+... | tri< lt _  _   =  contradiction lt x≮y
+... | tri≈ _  eq _   =  ≤-reflexive (sym eq)
+... | tri> _  _  y<x =  <⇒≤ {y} {x} y<x
+
+≰⇒> :  _≰_ ⇒ _>_
+≰⇒> {x} {y} x≰y  with <-cmp x y
+... | tri< lt _  _   =  contradiction (<⇒≤ {x} {y} lt) x≰y
+... | tri≈ _  eq _   =  contradiction (≤-reflexive eq) x≰y
+... | tri> _  _  x>y =  x>y
+
+≤,≢⇒< :  ∀ {x y} → x ≤ y → x ≢ y → x < y
+≤,≢⇒< (inj₁ x<y) _   =  x<y
+≤,≢⇒< (inj₂ x≡y) x≢y =  contradiction x≡y x≢y
+
+<-≤-trans :  ∀ {x y z} → x < y → y ≤ z → x < z
+<-≤-trans             x<y (inj₁ y<z) =  <-trans x<y y<z
+<-≤-trans {x} {_} {_} x<y (inj₂ y≡z) =  subst (x <_) y≡z x<y
+
+≤-<-trans :  ∀ {x y z} → x ≤ y → y < z → x < z
+≤-<-trans             (inj₁ x<y) y<z =  <-trans x<y y<z
+≤-<-trans {_} {_} {z} (inj₂ x≡y) y<z =  subst (_< z) (sym x≡y) y<z
+
+≡0-by≤ :  ∀ {x y} → x ≤ y → y ≡ zero → x ≡ zero
+≡0-by≤ {x} {y} x≤y y≡0 =  ≤0⇒≡ x≤0
+  where
+  x≤0 = subst (x ≤_) y≡0 x≤y
+
+≢0-by≤ :  ∀ {x y} → x ≤ y → x ≢ zero → y ≢ zero
+≢0-by≤ x≤y x≢0 =  x≢0 ∘ ≡0-by≤ x≤y
+
+_<?_ : Decidable _<_
+x <? y  with <-cmp x y
+... | tri< lt  _ _ =  yes lt
+... | tri≈ ¬lt _ _ =  no ¬lt
+... | tri> ¬lt _ _ =  no ¬lt
+
+_≤?_ : Decidable _≤_
+x ≤? y  with <-cmp x y
+... | tri< lt _  _  =  yes (<⇒≤ {x} {y} lt)
+... | tri≈ _  eq _  =  yes (≤-reflexive eq)
+... | tri> _  _  gt =  no (<⇒≱ {y} {x} gt)
+
+≤-isPreorder :  IsPreorder _≡_ _≤_
+≤-isPreorder =  record
+  { isEquivalence = isEquivalence
+  ; reflexive     = ≤-reflexive
+  ; trans         = ≤-trans
+  }
+
+<-resp-≈ :  _<_ Respects₂ _≡_
+<-resp-≈ =  ( (\{x y y'} y≡y' x<y → subst (x <_) y≡y' x<y) ,
```

Comment:
Can you prove the left and right versions separately? (If we keep the definition of `_<_`)

Reply:
Done